src/widgets/scrollbar.rs

```diff
@@ -498,25 +498,29 @@ impl Scrollbar<'_> {
     ///
     /// This method returns the length of the start, thumb, and end as a tuple.
     fn part_lengths(&self, area: Rect, state: &mut ScrollbarState) -> (usize, usize, usize) {
-        let track_len = self.track_length_excluding_arrow_heads(area) as f64;
-        let viewport_len = self.viewport_length(state, area) as f64;
+        let track_len = self.track_length_excluding_arrow_heads(area) as usize;
+        let viewport_len = self.viewport_length(state, area) as usize;
 
-        let content_length = state.content_length as f64;
-        // Clamp the position to show at least one line of the content, even if the content is
-        let position = state.position.min(state.content_length - 1) as f64;
+        // Clamp the position to show at least one line of the content
+        // Note: state.content_length is != 0 because otherwise render() returns immediately
+        let position = state.position.min(state.content_length - 1);
 
         // vscode style scrolling behavior (allow scrolling past end of content)
-        let scrollable_content_len = content_length + viewport_len - 1.0;
-        let thumb_start = position * track_len / scrollable_content_len;
-        let thumb_end = (position + viewport_len) * track_len / scrollable_content_len;
-
-        // We round just the positions (instead of floor / ceil), and then calculate the sizes from
-        // those positions. Rounding the sizes instead causes subtle off by 1 errors.
-        let track_start_len = thumb_start.round() as usize;
-        let thumb_end = thumb_end.round() as usize;
-
-        let thumb_len = thumb_end.saturating_sub(track_start_len);
-        let track_end_len = track_len as usize - track_start_len - thumb_len;
+        let scrollable_content_len = state.content_length + viewport_len - 1;
+
+        // Calculate the thumb start (inclusive) and end (exclusive) positions, with rounding
+        // and ensuring there is at least one visible thumb character.
+        let thumb_start =
+            (position * track_len + scrollable_content_len / 2) / scrollable_content_len;
```

Comment:
> Does adding .max(1) to line 518 to clamp the length to at least 1 solve this problem more clearly, or is there an edge case that this misses?

That won't be enough because we also need to ensure `thumb_start` is not right at the end of the track. Also, another reason why I converted to integer arithmetic is because in my experience this is less prone to weirdness.

> This code seems a bit "magical" in the sense that while it does exactly what is needed, it doesn't explain why this is necessary or how it benefits the problem. This makes it difficult to maintain in the future.

I've made the code more readable as of commit [1af36d9](https://github.com/ratatui-org/ratatui/pull/959/commits/1af36d9957ea8c389e3349615221f4d02679c408)